patches/server/0936-Add-Entity-hidden-by-default-flag.patch

```diff
@@ -0,0 +1,207 @@
+         return !this.getHandle().level.noCollision(this.getHandle(), aabb);
+     }
+     // Paper End - Collision API
++
++    // Paper start - Entity hiddenByDefault
++    @Override
++    public void setHiddenByDefault(boolean hiddenByDefault) {
++        this.getHandle().setHiddenByDefault(hiddenByDefault);
++
++        // We need to update the players again
++        ServerLevel world = ((CraftWorld) this.getWorld()).getHandle();
++        ChunkMap.TrackedEntity entityTracker = world.getChunkSource().chunkMap.entityMap.get(this.getEntityId());
++
++        // If the entity is already hiddenByDefault, we reset our "show to" set to again hide the entity from all players that may be seeing the entity
++        // We also clear the "show to" set when we stop hiding by default
++        entityTracker.showToEvenWhenHiddenByDefault.clear();
++
++        entityTracker.updatePlayers(this.server.getHandle().getPlayers());
++    }
++
++    @Override
++    public boolean isHiddenByDefault() {
++        return this.getHandle().isHiddenByDefault();
++    }
++    // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 067d64ce06f626ce90b3ccf608d98f3c64fce335..99ec84ebf018c36ddaf866b1ec05e64cb0f84b43 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -1721,6 +1721,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         ChunkMap tracker = ((ServerLevel) this.getHandle().level).getChunkSource().chunkMap;
+         ChunkMap.TrackedEntity entry = tracker.entityMap.get(other.getId());
+         if (entry != null) {
++            entry.showToEvenWhenHiddenByDefault.remove(this.getHandle().connection); // Paper - Entity hiddenByDefault
+             entry.removePlayer(this.getHandle());
+         }
+ 
+@@ -1758,13 +1759,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         if (this.equals(entity)) return;
+ 
+         Set<WeakReference<Plugin>> hidingPlugins = this.hiddenEntities.get(entity.getUniqueId());
+-        if (hidingPlugins == null) {
+-            return; // Entity isn't hidden
+-        }
+-        hidingPlugins.remove(CraftPlayer.getPluginWeakReference(plugin));
+-        if (!hidingPlugins.isEmpty()) {
+-            return; // Some other plugins still want the entity hidden
++        // Paper start - Entity hiddenByDefault
++        if (hidingPlugins != null) {
++            hidingPlugins.remove(CraftPlayer.getPluginWeakReference(plugin));
++            if (!hidingPlugins.isEmpty()) {
++                return; // Some other plugins still want the entity hidden
++            }
++        } else {
++            if (!entity.isHiddenByDefault()) {
++                return; // Entity isn't hidden
++            }
+         }
++        // Paper end
```

Comment:
Technically speaking here, we should also not re-register the shown entity if the player is already part of the default hidden overwrite set. While it does not make a difference in testing, as the client just discards a new ClientboundPlayerInfoPacket with the exact same profile, the method should not be called if the player can already see the entity.

Reply:
Ah! So if showPlayer is called twice on an entity hidden by default. Good catch!